Rename Snapshot to commit, move commit to branch api.

erickt opened the pull request:
Now to commit, you POST to either:

```
/api/projects/<projectId>/current-branch/commit
/api/projects/<projectId>/branches/<branchName>/commit
```

It also fixed a bug where creating a graph child wasn't getting a project link.

Commit:
Rename Snapshot to commit, move commit to branch api.

Now to commit, you POST to either:

/api/projects/<projectId>/current-branch/commit
/api/projects/<projectId>/branches/<branchName>/commit

Commit:
Fix graph children not getting the project link

cglewis commented:
Is there a concept of branching?  Or is this name change in anticipation of that?

erickt commented:
@cglewis: Yep, branches are already in. No merging though.

cglewis commented:
so what is the difference between branching and committing in this context?

erickt commented:
@cglewis: I guess it's a little confusing since we're using `branch` as both a noun and a verb. The branch action creates a branch and points it at an already existing graph. The commit action creates a cloned graph and updates an already existing branch to point at it.

cglewis commented:
@erickt that is confusing, as a branch implies that it is split off and separate, a.k.a. a clone, however in this instance it is not that.  Whereas a commit usually is thought of as the diff of changes since the previous commit (or original graph if no previous commits exist), however in this instance we're saying a complete clone of the original graph.  I think we should either change our verbiage to agree with the well-known definitions in the industry, or document this differentiation very well.

Commit:
Save the generated graph config in the GraphMetadata

Commit:
Rewrite DendriteGraph allow for setting to read-only

This also removes the implementation of `TitanGraph` from `DendriteGraph`.
This helps to keep our use of Titan transactionally safe.

Commit:
Fix UI accessing the ES backend

Commit:
Upgrade ng-upload to fix uploading on firefox 26

Commit:
Fix a thrown exception with DendriteRexsterGraph

Commit:
Minor cleanup

Commit:
Graph import does a commit for us.

Commit:
Fix checkpointing and opening graphs

Commit:
Fix id collision in project-detail.html

Commit:
Fix the histogram

Commit:
Fix betweenness UI

Commit:
Add vertexId to all vertices

Commit:
Fix rexster exception

Commit:
Index analytics in ES so the UI can display the fields

Commit:
Fix running groovy scripts

Commit:
Subclass DendriteGraph and DendriteGraphTx from TitanBlueprints*

Commit:
Update bin/dendrite to allow for commits

Commit:
getVertex should return null if the vertes doesn't exist

Commit:
Actually return a transaction

Commit:
bin/dendrite should trace the commit progress

Commit:
Move graph creation of a commit into the job thread.

This speeds up the appearance of creating a commit.

Commit:
Creating a branch should create a commit automatically

Commit:
Many branches can point at the same graph

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/DendriteGraph.java:
```diff
@@ -1,193 +1,176 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
-public class DendriteGraph implements TitanGraph {
+public class DendriteGraph extends TitanBlueprintsGraph {
```
What are the implications of this?

erickt replied:
It turns out it needs it to work with Rexster. Somewhere in that code base it does a cast of the graph to `TitanBlueprintsGraph`. It also lets us cut a lot of code out of that class.

cglewis replied:
Got it. Makes sense.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("name") == null) {
+                tx.makeKey("name")
+                        .dataType(String.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("type") == null) {
+                tx.makeKey("type")
+                        .dataType(String.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+                tx.makeKey("typeAndName")
+                        .dataType(String.class)
+                        .unique()
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("properties") == null) {
+                tx.makeKey("properties")
+                        .dataType(Properties.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("state") == null) {
+                tx.makeKey("state")
+                        .dataType(String.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("progress") == null) {
+                tx.makeKey("progress")
+                        .dataType(Float.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/metagraph/MetaGraph.java:
```diff
@@ -152,93 +151,100 @@ public void stop() {
+            if (tx.getType("mapreduceJobId") == null) {
+                tx.makeKey("mapreduceJobId")
+                        .dataType(String.class)
+                        .indexed(Vertex.class)
```
This needs "search" added to it.

cglewis commented on src/main/java/org/lab41/dendrite/web/controller/GraphImportController.java:
```diff
@@ -51,6 +50,8 @@
 
     static Logger logger = LoggerFactory.getLogger(GraphImportController.class);
 
+    static List<String> RESERVED_KEYS = Arrays.asList("id", "_id");
```
These are both reserved?  Is one for indexing and one for Titan, or why are both necessary?

erickt replied:
"id" and "label" are reserved by titan, "_id" is reserved by ES. I'll submit another patch to add "label".

cglewis replied:
Cool, sounds good.

cglewis commented on src/main/java/org/lab41/dendrite/web/controller/GraphImportController.java:
```diff
@@ -82,46 +83,52 @@
+                    if (tx.getType(key) == null && !RESERVED_KEYS.contains(key)) {
+                        tx.makeKey(key)
+                                .dataType(String.class)
+                                .indexed(Vertex.class)
```
This needs "search" added to it.

erickt replied:
it's indexed in elasticsearch one line down :)

cglewis replied:
oh, right. my bad.

erickt commented:
@cglewis: nice catch with the "search", I'll add it.
